ext/nokogiri/libxml2_polyfill.c

```diff
@@ -112,3 +112,14 @@ xmlSwitchEncodingName(xmlParserCtxtPtr ctxt, const char *encoding)
   return (xmlSwitchToEncoding(ctxt, handler));
 }
 #endif
+
+#ifndef HAVE_XMLADDIDSAFE
+int
+xmlAddIDSafe(xmlAttrPtr attr, const xmlChar *value) {
+  xmlIDPtr id = xmlAddID(NULL, attr->doc, value, attr);
```

Comment:
This can also fail and return NULL due to allocation failure. I think that in this case that will cause `xmlAddIDSafe` to return that the ID already exists.

Reply:
This one we can't do anything about. If we're using this polyfill then we're using libxml2 2.12 or earlier which doesn't have gnome/libxml2@e34a49b7 to return a separate error code if a memory allocation fails.

Reply:
This may not be worth dealing with but if `xmlAddID()` returns `NULL`, then we could do something like
```c
xmlError *lastError = xmlGetLastError();

if (lastError && lastError->code == XML_ERR_NO_MEMORY && lastError->level = XML_ERR_FATAL) {
    // Handle the OOM condition
}
```
That's from looking at https://github.com/GNOME/libxml2/blob/068abc0b75995c3d1576c078f01c76cc0814b280/valid.c#L56